File: src/config.test.ts
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { defaultConfig, load } from './config';

describe('defaultConfig', () => {
  it('should have non-empty output_dir', () => {
    const cfg = defaultConfig();
    expect(cfg.output_dir).toBeTruthy();
  });

  it('should have non-empty watches', () => {
    const cfg = defaultConfig();
    expect(cfg.watch.length).toBeGreaterThan(0);
    for (const w of cfg.watch) {
      expect(w.tool).toBeTruthy();
    }
  });
});

describe('load', () => {
  it('should create default config file if absent', () => {
    const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'histd-test-'));
    const cfgPath = path.join(dir, 'config.toml');

    const cfg = load(cfgPath);
    expect(cfg.output_dir).toBeTruthy();
    expect(fs.existsSync(cfgPath)).toBe(true);

    fs.rmSync(dir, { recursive: true });
  });

  it('should round-trip: second load reads the file created by first load', () => {
    const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'histd-test-'));
    const cfgPath = path.join(dir, 'config.toml');

    const first = load(cfgPath);
    const second = load(cfgPath);

    expect(first.output_dir).toBe(second.output_dir);

    fs.rmSync(dir, { recursive: true });
  });
});

File: internal/config/config_test.go
package config_test

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/inevolin/histd/internal/config"
)

func TestDefaultConfig(t *testing.T) {
	cfg, err := config.DefaultConfig()
	if err != nil {
		t.Fatalf("DefaultConfig() error: %v", err)
	}
	if cfg.OutputDir == "" {
		t.Error("OutputDir should not be empty")
	}
	if len(cfg.Watches) == 0 {
		t.Error("Watches should not be empty")
	}
	for _, w := range cfg.Watches {
		if w.Tool == "" {
			t.Errorf("watch entry %q has empty tool", w.Path)
		}
	}
}

func TestLoadCreatesDefaultFile(t *testing.T) {
	dir := t.TempDir()
	cfgPath := filepath.Join(dir, "config.toml")

	cfg, err := config.Load(cfgPath)
	if err != nil {
		t.Fatalf("Load() error: %v", err)
	}
	if cfg.OutputDir == "" {
		t.Error("OutputDir should not be empty after creating default config")
	}
	if _, err := os.Stat(cfgPath); err != nil {
		t.Errorf("config file was not created at %s: %v", cfgPath, err)
	}
}

func TestLoadRoundTrip(t *testing.T) {
	dir := t.TempDir()
	cfgPath := filepath.Join(dir, "config.toml")

	// First load creates the file.
	first, err := config.Load(cfgPath)
	if err != nil {
		t.Fatalf("first Load() error: %v", err)
	}

	// Second load reads the file that was created.
	second, err := config.Load(cfgPath)
	if err != nil {
		t.Fatalf("second Load() error: %v", err)
	}
	if first.OutputDir != second.OutputDir {
		t.Errorf("OutputDir mismatch: %q vs %q", first.OutputDir, second.OutputDir)
	}
}
